### PR TITLE
Cleanup generate rust

### PR DESCRIPTION
**Pull request by cramertj**
cc @qmx @Centril 

This is my first pass (still pretty rough) at reorganizing some of the code in src/generate/rust.rs in an attempt to make it a little easier for me to understand what's going on.

### CODE REVIEW

File: src/generate/rust.rs [hunk 1]
```diff
@@ -235,787 +257,88 @@ impl<Pat: Ord + Hash + MatchesEmpty + RustInputPat> Grammar<Pat> {
...
         while i < parse_nodes.borrow().len() {
             let rule = parse_nodes.borrow().get_index(i).unwrap().0.clone();
             rule.fill_parse_node_shape(&parse_nodes);
             i += 1;
         }
-        let mut all_parse_nodes: Vec<_> = named_parse_nodes.into_iter().map(|(k, s)| {
-            (k.clone(), s, Some(format!("{}<_>", k.0)))
+        let mut all_parse_nodes: Vec<ParseNode> = named_parse_nodes.into_iter().map(|(k, s)| {
+            ParseNode { kind: k.clone(), shape: s, ty: Some(format!("{}<_>", k.0)) }
         }).chain(parse_nodes.into_inner().into_iter().map(|(r, (k, s))| {
```
Comment (Centril) [inline src/generate/rust.rs]:
```suggestion
        }).chain(parse_nodes.into_inner().into_iter().map(|(r, (kind, s))| {
```

File: src/generate/rust.rs [hunk 2]
```diff
@@ -235,787 +257,88 @@ impl<Pat: Ord + Hash + MatchesEmpty + RustInputPat> Grammar<Pat> {
...
-            (k, s.unwrap(), match &*r {
-                Rule::RepeatMany(rule, _) | Rule::RepeatMore(rule, _) => match &**rule {
-                    Rule::Eat(_) => Some("[()]".to_string()),
-                    Rule::Call(r) => Some(format!("[{}<_>]", r)),
+            ParseNode {
+                kind: k,
```
Comment (Centril) [inline src/generate/rust.rs]:
```suggestion
                kind,
```

File: src/generate/rust.rs [hunk 3]
```diff
@@ -1481,12 +804,549 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
...
+                    } else {
+                        Err(ParseError::TooShort(handle))
+                    }
+                }))
+            })
+        }
+    }")
+}
+
+fn declare_rule<Pat>(
```
Comment (Centril) [inline src/generate/rust.rs:855]: This function is yuuge... (^,-) would be nice to split it up into more units... :)

**Review by Centril — COMMENTED** [review]
I only skimmed the changes... nice work :)

File: src/generate/rust.rs (hunk 3 above)
Comment (cramertj) [reply src/generate/rust.rs:855]: Done.

**Comment by cramertj** [comment]
Oh @qmx sorry I forgot to ping you yesterday after I fixed @Centril 's nits, but it's done now.

File: src/generate/rust.rs [hunk 4]
```diff
@@ -741,6 +742,7 @@ impl<'a, 'i, I: ::gll::runtime::Input> Handle<'a, 'i, I, ", name, "<'a, 'i, I>>
         }))().map_err(|::gll::runtime::MoreThanOne| Ambiguity(self))
     }
     pub fn all(self) -> impl Iterator<Item = ", name, "<'a, 'i, I>> {
+        #[allow(unused_variables)]
         let sppf = &self.parser.sppf;
```
Comment (eddyb) [inline src/generate/rust.rs]: What we've done in the past is rename `sppf` to `_sppf`.

File: src/generate/rust.rs [hunk 5]
```diff
@@ -464,7 +464,7 @@ pub enum ", name, "<'a, 'i: 'a, I: 'a + ::gll::runtime::Input> {");
 }");
             } else {
                 put!("
-
+#[allow(non_camel_case_types)]
```
Comment (eddyb) [inline src/generate/rust.rs]: What does this arise for?

File: src/generate/rust.rs [hunk 6]
```diff
@@ -716,6 +716,7 @@ impl<'a, 'i, I: ::gll::runtime::Input> Handle<'a, 'i, I, ", name, "<'a, 'i, I>>
     pub fn one(self) -> Result<", name, "<'a, 'i, I>, Ambiguity<Self>> {
         // HACK(eddyb) using a closure to catch `Err`s from `?`
         (|| Ok({
+            #[allow(unused_variables)]
             let sppf = &self.parser.sppf;
```
Comment (eddyb) [inline src/generate/rust.rs]: This would also be `_sppf`.

File: src/generate/rust.rs [hunk 7]
```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
         }
     }
 }
+
+fn impl_parse_with(name: &str, rust_slice_ty: &str, out: &mut String) {
+    put!(out, "
+    impl<'a, 'i, I: ::gll::runtime::Input<Slice = ", rust_slice_ty, ">> ", name, "<'a, 'i, I> {
```
Comment (eddyb) [inline src/generate/rust.rs:815]: Did this get indented more? They were intentionally set up to match up with the resulting file, indentation-wise.

File: src/generate/rust.rs [hunk 8]
```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
...
+                    if refutable {
+                        put!(out, "Option<");
+                    }
+                    put!(out, "Handle<'a, 'i, I, ", rule.field_pathset_type(paths), ">");
+                    if refutable {
+                        put!(out, ">");
+                    }
+                    put!(out, ",");
+                }
+                put!(out, "},");
```
Comment (eddyb) [inline src/generate/rust.rs:887]: `out` is repeated a lot, this is why I really liked the macro without it. In fact, I'd rather call it `p!` now or something short like that.

File: src/generate/rust.rs [hunk 9]
```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
...
+                        assert_eq!(paths.len(), 1);
+                        for p in paths.get_index(0).unwrap() {
+                            put!(out, " .", p);
+                        }
+                        put!(out, ", parser, _marker: PhantomData, },");
+                    }
+                }
+                put!(out, "}"); // -- end variant
+            }
+            put!(out, "}"); // -- end fn
```
Comment (eddyb) [inline src/generate/rust.rs:972]: Should we just move to quasi-quotation? Sadly @dtolnay is opposed to `${...}`-style nesting in quasi-quotation, which is the only way this code would look reasonable at all.

File: src/generate/rust.rs [hunk 10]
```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
...
+            put!(out, "}");
+        } else {
+            put!(out, "
+                let r = traverse!(one(sppf, node) ", rule.rule.generate_traverse_shape(false, &parse_nodes), ");
+                ", name, "::from_sppf(self.parser, node, r)");
+        }
+    }
+    put!(out, "
+        }))().map_err(|::gll::runtime::MoreThanOne| Ambiguity(self))
+        }
```
Comment (eddyb) [inline src/generate/rust.rs]: Yeah, indentation was screwed up during the move - it matters much less nowadays, as we're letting `rustfmt` clean up the output (if available).

Maybe we should just remove `#[cfg_attr(rustfmt, rustfmt_skip)]` and/or switch to quasi-quotation where possible (unbalanced `{}`/`[]`/`()` being the only issue with just using `stringify!`).

File: src/generate/templates/header.rs [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+pub type ParseResult<'a, 'i, I, T> =
+    Result<Handle<'a, 'i, I, T>, ParseError<Handle<'a, 'i, I, T>>>;
+
+pub type Any = dyn any::Any;
+
+#[derive(Debug)]
+pub struct Ambiguity<T>(T);
+
+pub struct Handle<'a, 'i: 'a, I: 'a + ::gll::runtime::Input, T: ?Sized> {
```
Comment (eddyb) [inline src/generate/templates/header.rs:15]: I kind of wish all of this stuff could be in the `gll` crate (or a `gll-runtime` crate) and parametrized by `_P` and `_C`, with local aliases.

I wonder if coherence would let us do it the tedious way.

But also, there's no reason we need both `_P` and `_C`, one of them (or a third type) could implement a trait to get the other.
Maybe `I` could be combined into this somehow? As `_P, _C, I` are the 3 type parameters of the `Parser`.

File: src/generate/rust.rs [hunk 12]
```diff
@@ -28,38 +28,58 @@ impl<S: AsRef<str>> RustInputPat for scannerless::Pat<S> {
     }
 }
 
+type ParseNodeMap<Pat> =
+    OrderMap<Rc<Rule<Pat>>, (ParseNodeKind, Option<ParseNodeShape<ParseNodeKind>>)>;
```
Comment (eddyb) [inline src/generate/rust.rs:32]: You could call this `ParseNodeCache`, maybe?

File: src/generate/rust.rs [hunk 13]
```diff
@@ -28,38 +28,58 @@ impl<S: AsRef<str>> RustInputPat for scannerless::Pat<S> {
...
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct ParseNode {
```
Comment (eddyb) [inline src/generate/rust.rs:35]: Maybe rename this to `TypedParseNode`?

File: src/generate/rust.rs [hunk 14]
```diff
@@ -28,38 +28,58 @@ impl<S: AsRef<str>> RustInputPat for scannerless::Pat<S> {
...
+}
+
+struct Variant<'a, Pat> {
+    rule: Rc<Rule<Pat>>,
+    name: &'a str,
+    fields: OrderMap<&'a str, OrderSet<Vec<usize>>>,
+}
+
 impl<Pat: PartialEq> RuleWithNamedFields<Pat> {
     fn find_variant_fields(
```
Comment (eddyb) [inline src/generate/rust.rs:48]: This should probably be `extract_variants`, especially with the better type.

File: src/generate/rust.rs [hunk 15]
```diff
@@ -279,10 +277,10 @@ impl<Pat: Ord + Hash + MatchesEmpty + RustInputPat> Grammar<Pat> {
         }
         let mut all_parse_nodes: Vec<ParseNode> = named_parse_nodes.into_iter().map(|(k, s)| {
             ParseNode { kind: k.clone(), shape: s, ty: Some(format!("{}<_>", k.0)) }
-        }).chain(parse_nodes.into_inner().into_iter().map(|(r, (k, s))| {
+        }).chain(parse_nodes.into_inner().into_iter().map(|(r, (kind, shape))| {
```
Comment (eddyb) [inline src/generate/rust.rs:280]: It would probably be nice to have a `struct` type for the cache entry here.

File: src/generate/rust.rs [hunk 16]
```diff
@@ -412,9 +410,7 @@ impl<'a> Continuation<'a> {
 
     #[cfg_attr(rustfmt, rustfmt_skip)]
     fn reify_as(&mut self, label: CodeLabel) {
-        let code = format!("
-            {} => {{{}
-            }}", label, self.to_inline());
+        let code = format!("{} => {{{}}}", label, self.to_inline());
```
Comment (eddyb) [inline src/generate/rust.rs:413]: This is horrible, aaaaaaaaaaah!!!
(not your fault, just `{{` and `}}` are bad escapes)
We should just have a version of `put!` that's like `format!` tbh.

File: src/generate/rust.rs [hunk 17]
```diff
@@ -913,10 +910,28 @@ where
...
 
     // Implement `Debug`;
-    rule_debug_impls(name, &rule, variants.as_ref().map(|x| &**x), out);
+    rule_debug_impls(name, &rule, variants, out);
+
+    impl_rule_from_sppf(name, &rule, variants, parse_nodes, out);
 
+    put!(out, "impl<'a, 'i, I: ::gll::runtime::Input> Handle<'a, 'i, I, ", name, "<'a, 'i, I>> {");
+    impl_rule_one(name, &rule, variants, parse_nodes, out);
+    impl_rule_all(name, &rule, variants, parse_nodes, out);
```
Comment (eddyb) [inline src/generate/rust.rs:919]: Can we just have a context with all of these fields on it?

**Review by eddyb — COMMENTED** [review]
Sorry for being so late to see this!

It seems nice overall, but I still left some comments.
If you don't have time and I do, I might get to them.

**Pull request merged**